impl/src/main/java/org/jboss/weld/util/Beans.java

```diff
@@ -640,4 +644,44 @@ public static boolean hasBuiltinScope(Bean<?> bean) {
         manager.getServices().get(SlimAnnotatedTypeStore.class).put(implementationClass.slim());
         return implementationClass;
     }
+
+    static ImmutableSet.Builder<Type> omitIllegalBeanTypes(Set<Type> types, EnhancedAnnotated<?, ?> annotated) {
+        ImmutableSet.Builder<Type> builder = ImmutableSet.builder();
+        for (Type type : types) {
+            if (isIllegalBeanType(type)) {
+                MetadataLogger.LOG.illegalBeanTypeIgnored(type, annotated);
+            } else {
+                builder.add(type);
+            }
+        }
+        return builder;
+    }
+
+    /**
+     *
+     * @param beanType
+     * @return <code>true</code> if the given type is not a legal bean type, <code>false</code> otherwise
+     */
+    public static boolean isIllegalBeanType(Type beanType) {
```

Comment:
`Types` - yes, makes sense
`Foo<Foo<?>>` - we already do this, see https://github.com/weld/core/pull/807/files#diff-f85978750c079061a382ea51cb405f96R64